src/compiler/checker.ts

```diff
@@ -11459,9 +11459,9 @@ namespace ts {
             }
         }
 
-        function inBlockOrObjectLiteralExpression(node: Node) {
+        function inObjectLiteralExpression(node: Node) {
             while (node) {
-                if (node.kind === SyntaxKind.Block || node.kind === SyntaxKind.ObjectLiteralExpression) {
+                if (node.kind === SyntaxKind.ObjectLiteralExpression) {

```

Comment:
This breaks for a class inside an object literal:
```ts
var obj = {
    class: class {
        static foo;
    }
}
```
Instead of doing a loop, I think you should just check the kind of the parent (or perhaps the grandparent, whichever it is).